tailscale.com/util/deephash fail on armhf & armv7

https://github.com/tailscale/tailscale/tree/release-branch/1.14, https://github.com/tailscale/tailscale/tree/v1.14.0 and https://github.com/tailscale/tailscale/tree/main all fail on `armhf` & `armv7` mostly on `TestArrayAllocs`
tested on Raspberry Pi 3B+ Raspbian 11 (bullseye) & Alpine in Docker + qemu
with Go 1.16.7 & 1.17
```
pi@bengal ~/tailscale (main)> ~/sdk/go1.17/bin/go test tailscale.com/util/deephash
--- FAIL: TestArrayAllocs (0.01s)
    deephash_test.go:343: allocs = 2; want 0
FAIL
FAIL    tailscale.com/util/deephash     1.004s
FAIL
```
additionally `tailscale.com/tstime/rate` randomly fails
```
--- FAIL: TestLongRunningQPS (5.04s)
    rate_test.go:205: numOK = 599, want 602 (ideal 603.006834)
FAIL
FAIL	tailscale.com/tstime/rate	8.104s
```
additional logs
https://gitlab.alpinelinux.org/panekj/aports/-/jobs/472227
https://gitlab.alpinelinux.org/panekj/aports/-/jobs/472226

Able to reproduce on `GOARCH=arm`, but not on `GOARCH=arm64`:
```
$ GOARCH=arm go test -c && qemu-arm deephash.test 
--- FAIL: TestArrayAllocs (0.01s)
    deephash_test.go:343: allocs = 2; want 0
FAIL

$ GOARCH=arm64 go test -c && qemu-aarch64 deephash.test 
PASS
```

--- a/util/deephash/deephash_test.go
+++ b/util/deephash/deephash_test.go
@@ -8,9 +8,11 @@ import (
 	"archive/tar"
 	"bufio"
 	"bytes"
+	"crypto/sha256"
 	"fmt"
 	"math"
 	"reflect"
+	"runtime"
 	"testing"
 
 	"inet.af/netaddr"
@@ -332,15 +334,32 @@ func TestArrayAllocs(t *testing.T) {
 	if version.IsRace() {
 		t.Skip("skipping test under race detector")
 	}
+
+	// In theory, there should be no allocations. However, escape analysis on
+	// certain architectures fails to detect that certain cases do not escape.
+	// This discrepency currently affects sha256.digest.Sum.
+	// Measure the number of allocations in sha256 to ensure that Hash does
+	// not allocate on top of its usage of sha256.
+	// See https://golang.org/issue/48055.
+	var b []byte
+	h := sha256.New()
+	want := int(testing.AllocsPerRun(1000, func() {
+		b = h.Sum(b[:0])
+	}))
+	switch runtime.GOARCH {
+	case "amd64", "arm64":
+		want = 0 // ensure no allocations on popular architectures
+	}
+
 	type T struct {
 		X [32]byte
 	}
 	x := &T{X: [32]byte{1: 1, 2: 2, 3: 3, 4: 4}}
-	n := int(testing.AllocsPerRun(1000, func() {
+	got := int(testing.AllocsPerRun(1000, func() {
 		sink = Hash(x)
 	}))
-	if n > 0 {
-		t.Errorf("allocs = %v; want 0", n)
+	if got > want {
+		t.Errorf("allocs = %v; want %v", got, want)
 	}
 }
 

